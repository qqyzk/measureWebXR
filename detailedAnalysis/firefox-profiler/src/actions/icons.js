/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

// @flow
import type { Action, ThunkAction } from 'firefox-profiler/types';

export function iconHasLoaded(icon: string): Action {
  return {
    type: 'ICON_HAS_LOADED',
    icon,
  };
}

export function iconIsInError(icon: string): Action {
  return {
    type: 'ICON_IN_ERROR',
    icon,
  };
}

const icons: Set<string> = new Set();

type IconRequestResult = 'loaded' | 'error' | 'cached';

function _getIcon(icon: string): Promise<IconRequestResult> {
  if (icons.has(icon)) {
    return Promise.resolve('cached');
  }

  icons.add(icon);

  const result = new Promise((resolve) => {
    const image = new Image();
    image.src = icon;
    image.referrerPolicy = 'no-referrer';
    image.onload = () => {
      resolve('loaded');
    };
    image.onerror = () => {
      resolve('error');
    };
  });

  return result;
}

export function iconStartLoading(icon: string): ThunkAction<Promise<void>> {
  return (dispatch) => {
    return _getIcon(icon).then((result) => {
      switch (result) {
        case 'loaded':
          dispatch(iconHasLoaded(icon));
          break;
        case 'error':
          dispatch(iconIsInError(icon));
          break;
        case 'cached':
          // nothing to do
          break;
        default:
          throw new Error(`Unknown icon load result ${result}`);
      }
    });
  };
}
