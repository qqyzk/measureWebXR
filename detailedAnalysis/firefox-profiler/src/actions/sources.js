/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

// @flow
import type { Action, SourceLoadingError } from 'firefox-profiler/types';

export function beginLoadingSourceFromUrl(file: string, url: string): Action {
  return { type: 'SOURCE_LOADING_BEGIN_URL', file, url };
}

export function beginLoadingSourceFromBrowserConnection(file: string): Action {
  return { type: 'SOURCE_LOADING_BEGIN_BROWSER_CONNECTION', file };
}

export function finishLoadingSource(file: string, source: string): Action {
  return { type: 'SOURCE_LOADING_SUCCESS', file, source };
}

export function failLoadingSource(
  file: string,
  errors: SourceLoadingError[]
): Action {
  return { type: 'SOURCE_LOADING_ERROR', file, errors };
}
