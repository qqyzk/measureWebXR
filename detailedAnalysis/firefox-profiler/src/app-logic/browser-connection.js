/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
// @flow

import { oneLine } from 'common-tags';
import {
  getProfileViaWebChannel,
  getSymbolTableViaWebChannel,
  querySupportsGetProfileAndSymbolicationViaWebChannel,
  querySymbolicationApiViaWebChannel,
} from './web-channel';

/**
 * This file manages the communication between the profiler and the browser.
 */

export type BrowserConnectionStatus =
  // The initial state.
  | {| status: 'NO_ATTEMPT' |}
  // In non-Firefox browsers we don't attempt to establish a connection.
  // This is determined via the userAgent.
  | {| status: 'NOT_FIREFOX' |}
  // We are in Firefox, and have sent the initial WebChannel event.
  | {| status: 'WAITING' |}
  // We are in Firefox but the WebChannel connection has been denied.
  // This usually means that this profiler instance is running on a
  // different host than the one that's specified in the
  // preference `devtools.performance.recording.ui-base-url`.
  | {| status: 'DENIED', error: Error |}
  // We are in Firefox but the WebChannel did not respond within 5 seconds.
  // This is unexpected. It could mean that we are running in an old Firefox
  // (older than Firefox 76) which did not have a profiler WebChannel.
  | {| status: 'TIMED_OUT' |}
  // The WebChannel connection has been established.
  | {| status: 'ESTABLISHED', browserConnection: BrowserConnection |};

/**
 * The interface of communication with the browser. Can be backed by a WebChannel
 * or by the frame script API.
 * Only exists if at least an old version of the WebChannel is available in this browser.
 */
export interface BrowserConnection {
  // Get the profile for this tab from the browser.
  getProfile(options: {|
    onThirtySecondTimeout: () => void,
  |}): Promise<ArrayBuffer | MixedObject>;

  // Query the browser-internal symbolication API. This provides richer
  // information than getSymbolTable.
  querySymbolicationApi(path: string, requestJson: string): Promise<string>;

  // Get a symbol table from the browser.
  getSymbolTable(
    debugName: string,
    breakpadId: string
  ): Promise<SymbolTableAsTuple>;
}

/**
 * The regular implementation of the BrowserConnection interface.
 *
 * Only created when a WebChannel exists. But it could be an old WebChannel
 * (from a pre-bug 1625309 Firefox version) which does not support obtaining
 * the profile or symbols. So this class also supports the frame script.
 */
class BrowserConnectionImpl implements BrowserConnection {
  _webChannelSupportsGetProfileAndSymbolication: boolean;
  _geckoProfiler: $GeckoProfiler | void;

  constructor(webChannelSupportsGetProfileAndSymbolication: boolean) {
    this._webChannelSupportsGetProfileAndSymbolication =
      webChannelSupportsGetProfileAndSymbolication;
  }

  // Only called when we must obtain the profile from the browser, i.e. if we
  // cannot proceed without a connection to the browser. This method falls back
  // to the frame script API (window.geckoProfilerPromise) if this browser has
  // an old version of the WebChannel.
  async _getConnectionViaFrameScript(): Promise<$GeckoProfiler> {
    if (!this._geckoProfiler) {
      this._geckoProfiler = await window.geckoProfilerPromise;
    }
    return this._geckoProfiler;
  }

  async getProfile(options: {|
    onThirtySecondTimeout: () => void,
  |}): Promise<ArrayBuffer | MixedObject> {
    const timeoutId = setTimeout(options.onThirtySecondTimeout, 30000);

    // On Firefox 96 and above, we can get the profile from the WebChannel.
    if (this._webChannelSupportsGetProfileAndSymbolication) {
      const profile = await getProfileViaWebChannel();
      clearTimeout(timeoutId);
      return profile;
    }

    // For older versions, fall back to the geckoProfiler frame script API.
    // This fallback can be removed once the oldest supported Firefox ESR version is 96 or newer.
    const geckoProfiler = await this._getConnectionViaFrameScript();
    const profile = await geckoProfiler.getProfile();
    clearTimeout(timeoutId);
    return profile;
  }

  async querySymbolicationApi(
    path: string,
    requestJson: string
  ): Promise<string> {
    // This only works on Firefox 96 and above.
    if (!this._webChannelSupportsGetProfileAndSymbolication) {
      throw new Error(
        "Can't use querySymbolicationApi in Firefox versions with the old WebChannel."
      );
    }

    return querySymbolicationApiViaWebChannel(path, requestJson);
  }

  async getSymbolTable(
    debugName: string,
    breakpadId: string
  ): Promise<SymbolTableAsTuple> {
    // On Firefox 96 and above, we can get the symbol table from the WebChannel.
    if (this._webChannelSupportsGetProfileAndSymbolication) {
      return getSymbolTableViaWebChannel(debugName, breakpadId);
    }

    // For older versions, fall back to the geckoProfiler frame script API.
    // This fallback can be removed once the oldest supported Firefox ESR version is 96 or newer.
    // Note that we use this._geckoProfiler directly instead of
    // _getConnectionViaFrameScript so that we're not waiting forever when the
    // user opens an unsymbolicated profile with a Firefox that doesn't support
    // the WebChannel.
    if (this._geckoProfiler) {
      return this._geckoProfiler.getSymbolTable(debugName, breakpadId);
    }

    throw new Error(
      'Cannot obtain a symbol table: have neither WebChannel nor a GeckoProfiler object'
    );
  }
}

function _isFirefox(userAgent: string): boolean {
  return Boolean(userAgent.match(/Firefox\/\d+\.\d+/));
}

class TimeoutError extends Error {
  name = 'TimeoutError';
}

function makeTimeoutRejectionPromise(durationInMs) {
  return new Promise((_resolve, reject) => {
    setTimeout(() => {
      reject(new TimeoutError(`Timed out after ${durationInMs}ms`));
    }, durationInMs);
  });
}

export async function createBrowserConnection(
  userAgent: string = navigator.userAgent
): Promise<BrowserConnectionStatus> {
  if (!_isFirefox(userAgent)) {
    return { status: 'NOT_FIREFOX' };
  }
  try {
    const webChannelSupportsGetProfileAndSymbolication = await Promise.race([
      querySupportsGetProfileAndSymbolicationViaWebChannel(),
      makeTimeoutRejectionPromise(5000),
    ]);
    // If we get here, it means querySupportsGetProfileAndSymbolicationViaWebChannel()
    // did not throw an exception. This means that a WebChannel exists.
    const browserConnection = new BrowserConnectionImpl(
      webChannelSupportsGetProfileAndSymbolication
    );
    return {
      status: 'ESTABLISHED',
      browserConnection,
    };
  } catch (e) {
    if (e instanceof TimeoutError) {
      // The browser never reacted to our WebChannelMessageToChrome event.
      // This can happen if we're running on a browser that's not Firefox, or if we're running
      // on an old version of Firefox which does not have support for any WebChannels.
      return { status: 'TIMED_OUT' };
    }
    // The WebChannel responded with an error. This usually means that this profiler
    // instance is running on a different host than the one that's specified in the
    // preference `devtools.performance.recording.ui-base-url`.
    // Or it means we're running in a test environment where no WebChannel simulation
    // has been set up.
    const error = new Error(oneLine`
       This profiler instance was unable to connect to the
       WebChannel. This usually means that itâ€™s running on a
       different host from the one that is specified in the
       preference devtools.performance.recording.ui-base-url. If
       you would like to capture new profiles with this instance, you can go to about:config
       and change the preference. Error: ${e.name}: ${e.message}
     `);
    return { status: 'DENIED', error };
  }
}
