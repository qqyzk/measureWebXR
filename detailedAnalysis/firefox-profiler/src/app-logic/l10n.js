/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

// @flow

import { FluentBundle, FluentResource } from '@fluent/bundle';
import {
  PSEUDO_STRATEGIES,
  PSEUDO_STRATEGIES_DIRECTION,
} from 'firefox-profiler/utils/l10n-pseudo';
import { SHORTDATE } from 'firefox-profiler/utils/l10n-ftl-functions';

// This object contains the locales that are at least 90% complete and
// that we enable by default. Don't forget to update the array RTL_LOCALES when
// adding a RTL locale, if necessary.
// The localized names are copied from the curated list that can be found at
// https://mozilla-l10n.github.io/firefox-languages/.
// We use the language names in their own language.
// Also note that the order specified here is the order they'll be displayed in the
// language switcher, so it's important to keep the alphabetical order.
export const AVAILABLE_LOCALES_TO_LOCALIZED_NAMES = {
  de: 'Deutsch',
  el: 'Ελληνικά',
  'en-GB': 'English (GB)',
  'en-US': 'English (US)',
  'es-CL': 'Español', // Use "Español (CL)" once we have more spanish versions
  fr: 'Français',
  'fy-NL': 'Frysk',
  ia: 'Interlingua',
  it: 'Italiano',
  nl: 'Nederlands',
  'pt-BR': 'Português (BR)',
  'sv-SE': 'Svenska',
  uk: 'Українська',
  'zh-CN': '简体中文',
  'zh-TW': '正體中文',
};

// This constant contains all locales available to our application. The default
// is to use the keys of the previous object.
// However when running the yarn scripts with the environment variable L10N=1,
// webpack replaces AVAILABLE_STAGING_LOCALES with all locales that have an FTL
// file in the repository, and this becomes the value for this constant. This
// is used in our l10n branch when deployed on netlify.
export const AVAILABLE_LOCALES: Array<string> =
  AVAILABLE_STAGING_LOCALES ||
  Object.keys(AVAILABLE_LOCALES_TO_LOCALIZED_NAMES);

export const DEFAULT_LOCALE = 'en-US';

// This array contains only the locales that are RTL (Right-To-Left).
// This list has been copied from the l20n library in the Firefox OS project
// (ancestor of Fluent):
// https://github.com/mozilla-b2g/gaia/blob/975a35c0f5010df341e96d6c5ec60217f5347412/shared/js/intl/l20n-client.js#L31-L35
const RTL_LOCALES = ['ar', 'he', 'fa', 'ps', 'ur'];

/**
 * Fetches ftl file of different locales.
 * Returns the locale and the ftl string grouped as an Array.
 */
export async function fetchMessages(locale: string): Promise<[string, string]> {
  const response = await fetch(`/locales/${locale}/app.ftl`);
  const messages = await response.text();
  return [locale, messages];
}

/**
 * A generator function responsible for building the sequence
 * of FluentBundle instances in the order of user's language
 * preferences.
 */
export function* lazilyParsedBundles(
  fetchedMessages: Array<[string, string]>,
  pseudoStrategy?: 'accented' | 'bidi' | null
): Generator<FluentBundle, void, void> {
  const transform = pseudoStrategy
    ? PSEUDO_STRATEGIES[pseudoStrategy]
    : undefined;
  for (const [locale, messages] of fetchedMessages) {
    const resource = new FluentResource(messages);
    const bundle = new FluentBundle(locale, {
      transform,
      functions: { SHORTDATE },
    });
    bundle.addResource(resource);
    yield bundle;
  }
}

/**
 * From a language string and a pseudoStrategy, this returns which direction the
 * document should be in.
 */
export function getLocaleDirection(
  language: string,
  pseudoStrategy?: 'accented' | 'bidi' | null
): 'ltr' | 'rtl' {
  if (pseudoStrategy && pseudoStrategy in PSEUDO_STRATEGIES_DIRECTION) {
    return PSEUDO_STRATEGIES_DIRECTION[pseudoStrategy];
  }

  // The language can be simple like "fr" but also more precise like "fr-FR".
  const tag = language.split('-')[0];
  return RTL_LOCALES.includes(tag) ? 'rtl' : 'ltr';
}
