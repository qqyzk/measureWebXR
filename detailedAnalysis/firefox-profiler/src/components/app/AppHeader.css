/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

.appHeader {
  display: flex;
  justify-content: space-between;
  padding-bottom: 0.5em;
  border-bottom: 1px solid #ddd;

  /* Negative margin is here to compensate for the children's margins */
  margin: 0 -0.2em 0.5em;
  font-size: 1.5em;
}

.appHeaderLink {
  color: inherit;
  text-decoration: none;
}

/* Note that :focus-visible styles are defined in focus.css */
.appHeaderLink:hover {
  text-decoration: underline;
}

.appHeaderSlogan {
  margin: 0 0.2em;
}

.appHeaderSubtext {
  font-weight: normal;
}

.appHeaderGithubIcon {
  margin: 0 0.2em;
  transition: opacity 250ms var(--animation-curve);
}

.appHeaderGithubIcon svg {
  vertical-align: middle;
}

.appHeaderGithubIcon:hover {
  opacity: 0.5;
}
