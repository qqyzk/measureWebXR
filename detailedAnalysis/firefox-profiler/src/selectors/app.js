/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

// @flow
import { createSelector } from 'reselect';

import {
  getDataSource,
  getSelectedTab,
  getTimelineTrackOrganization,
  getHiddenGlobalTracks,
  getHiddenLocalTracksByPid,
  getIsActiveTabResourcesPanelOpen,
} from './url-state';
import {
  getGlobalTracks,
  getLocalTracksByPid,
  getActiveTabTimeline,
} from './profile';
import { getZipFileState } from './zipped-profiles.js';
import { assertExhaustiveCheck, ensureExists } from '../utils/flow';
import {
  FULL_TRACK_SCREENSHOT_HEIGHT,
  ACTIVE_TAB_TRACK_SCREENSHOT_HEIGHT,
  TRACK_NETWORK_HEIGHT,
  TRACK_MEMORY_HEIGHT,
  TRACK_IPC_HEIGHT,
  TRACK_PROCESS_BLANK_HEIGHT,
  TIMELINE_RULER_HEIGHT,
  TRACK_VISUAL_PROGRESS_HEIGHT,
  ACTIVE_TAB_TIMELINE_RESOURCES_HEADER_HEIGHT,
  TRACK_EVENT_DELAY_HEIGHT,
  TIMELINE_MARGIN_LEFT,
  ACTIVE_TAB_TIMELINE_MARGIN_LEFT,
  TRACK_PROCESS_CPU_HEIGHT,
} from '../app-logic/constants';

import type {
  AppState,
  AppViewState,
  UrlSetupPhase,
  Selector,
  CssPixels,
  ThreadsKey,
  ExperimentalFlags,
  UploadedProfileInformation,
} from 'firefox-profiler/types';
import type { TabSlug } from 'firefox-profiler/app-logic/tabs-handling';
import type {
  BrowserConnectionStatus,
  BrowserConnection,
} from 'firefox-profiler/app-logic/browser-connection';

/**
 * Simple selectors into the app state.
 */
export const getApp: Selector<AppState> = (state) => state.app;
export const getView: Selector<AppViewState> = (state) => getApp(state).view;
export const getUrlSetupPhase: Selector<UrlSetupPhase> = (state) =>
  getApp(state).urlSetupPhase;
export const getHasZoomedViaMousewheel: Selector<boolean> = (state) => {
  return getApp(state).hasZoomedViaMousewheel;
};
export const getIsSidebarOpen: Selector<boolean> = (state) =>
  getApp(state).isSidebarOpenPerPanel[getSelectedTab(state)];
export const getPanelLayoutGeneration: Selector<number> = (state) =>
  getApp(state).panelLayoutGeneration;
export const getLastVisibleThreadTabSlug: Selector<TabSlug> = (state) =>
  getApp(state).lastVisibleThreadTabSlug;
export const getTrackThreadHeights: Selector<{
  [key: ThreadsKey]: CssPixels,
}> = (state) => getApp(state).trackThreadHeights;
export const getIsNewlyPublished: Selector<boolean> = (state) =>
  getApp(state).isNewlyPublished;
export const getExperimental: Selector<ExperimentalFlags> = (state) =>
  getApp(state).experimental;
export const getBrowserConnectionStatus: Selector<BrowserConnectionStatus> = (
  state
) => getApp(state).browserConnectionStatus;
export const getBrowserConnection: Selector<BrowserConnection | null> =
  createSelector(getBrowserConnectionStatus, (status) =>
    status.status === 'ESTABLISHED' ? status.browserConnection : null
  );
export const getIsEventDelayTracksEnabled: Selector<boolean> = (state) =>
  getExperimental(state).eventDelayTracks;
export const getIsExperimentalCPUGraphsEnabled: Selector<boolean> = (state) =>
  getExperimental(state).cpuGraphs;
export const getIsExperimentalProcessCPUTracksEnabled: Selector<boolean> = (
  state
) => getExperimental(state).processCPUTracks;

export const getIsDragAndDropDragging: Selector<boolean> = (state) =>
  getApp(state).isDragAndDropDragging;
export const getIsDragAndDropOverlayRegistered: Selector<boolean> = (state) =>
  getApp(state).isDragAndDropOverlayRegistered;

export const getCurrentProfileUploadedInformation: Selector<
  UploadedProfileInformation | null
> = (state) => getApp(state).currentProfileUploadedInformation;

/**
 * Height of screenshot track is different depending on the view.
 */
export const getScreenshotTrackHeight: Selector<number> = createSelector(
  getTimelineTrackOrganization,
  (timelineTrackOrganization) => {
    switch (timelineTrackOrganization.type) {
      case 'active-tab':
        return ACTIVE_TAB_TRACK_SCREENSHOT_HEIGHT;
      case 'full':
      case 'origins':
        return FULL_TRACK_SCREENSHOT_HEIGHT;
      default:
        throw assertExhaustiveCheck(
          timelineTrackOrganization,
          `Unhandled TimelineTrackOrganization`
        );
    }
  }
);

/**
 * This selector takes all of the tracks, and deduces the height in CssPixels
 * of the timeline. This is here to calculate the max-height of the timeline
 * for the splitter component.
 *
 * The height of the component is determined by the sizing of each track in the list.
 * Most sizes are pretty static, and are set through values in the component. The only
 * tricky value to determine is the thread track. These values get reported to the store
 * and get added in here.
 */
export const getTimelineHeight: Selector<null | CssPixels> = createSelector(
  getTimelineTrackOrganization,
  getGlobalTracks,
  getLocalTracksByPid,
  getHiddenGlobalTracks,
  getHiddenLocalTracksByPid,
  getTrackThreadHeights,
  getActiveTabTimeline,
  getIsActiveTabResourcesPanelOpen,
  getScreenshotTrackHeight,
  (
    timelineTrackOrganization,
    globalTracks,
    localTracksByPid,
    hiddenGlobalTracks,
    hiddenLocalTracksByPid,
    trackThreadHeights,
    activeTabTimeline,
    isActiveTabResourcesPanelOpen,
    screenshotTrackHeight
  ) => {
    let height = TIMELINE_RULER_HEIGHT;
    const border = 1;
    switch (timelineTrackOrganization.type) {
      case 'origins': {
        return height + 500;
      }
      case 'active-tab': {
        if (activeTabTimeline.resources.length > 0) {
          // Active tab resources panel has a header and we should also add its
          // height if there is a panel there.
          height += ACTIVE_TAB_TIMELINE_RESOURCES_HEADER_HEIGHT;
        }

        // Add the height of the main track.
        // The thread tracks have enough complexity that it warrants measuring
        // them rather than statically using a value like the other tracks.
        const { threadsKey } = activeTabTimeline.mainTrack;
        const trackThreadHeight = trackThreadHeights[threadsKey];
        if (trackThreadHeight === undefined) {
          // The height isn't computed yet, return.
          return null;
        }
        height += trackThreadHeight + border;

        // Add the height of screenshot tracks.
        for (let i = 0; i < activeTabTimeline.screenshots.length; i++) {
          height += screenshotTrackHeight + border;
        }

        if (isActiveTabResourcesPanelOpen) {
          for (const resourceTrack of activeTabTimeline.resources) {
            switch (resourceTrack.type) {
              case 'sub-frame':
              case 'thread':
                {
                  // The thread tracks have enough complexity that it warrants measuring
                  // them rather than statically using a value like the other tracks.
                  const { threadIndex } = resourceTrack;
                  if (threadIndex === null) {
                    height += TRACK_PROCESS_BLANK_HEIGHT + border;
                  } else {
                    const trackThreadHeight = trackThreadHeights[threadIndex];
                    if (trackThreadHeight === undefined) {
                      // The height isn't computed yet, return.
                      return null;
                    }
                    height +=
                      trackThreadHeight +
                      ACTIVE_TAB_TIMELINE_RESOURCES_HEADER_HEIGHT +
                      border;
                  }
                }
                break;
              default:
                throw assertExhaustiveCheck(resourceTrack);
            }
          }
        }

        return height;
      }
      case 'full': {
        for (const [trackIndex, globalTrack] of globalTracks.entries()) {
          if (!hiddenGlobalTracks.has(trackIndex)) {
            switch (globalTrack.type) {
              case 'screenshots':
                height += screenshotTrackHeight + border;
                break;
              case 'visual-progress':
              case 'perceptual-visual-progress':
              case 'contentful-visual-progress':
                height += TRACK_VISUAL_PROGRESS_HEIGHT;
                break;
              case 'process': {
                // The thread tracks have enough complexity that it warrants measuring
                // them rather than statically using a value like the other tracks.
                const { mainThreadIndex } = globalTrack;
                if (mainThreadIndex === null) {
                  height += TRACK_PROCESS_BLANK_HEIGHT + border;
                } else {
                  const trackThreadHeight = trackThreadHeights[mainThreadIndex];
                  if (trackThreadHeight === undefined) {
                    // The height isn't computed yet, return.
                    return null;
                  }
                  height += trackThreadHeight + border;
                }
                break;
              }
              default:
                throw assertExhaustiveCheck(globalTrack);
            }
          }
        }

        // Figure out which PIDs are hidden.
        const hiddenPids = new Set();
        for (const trackIndex of hiddenGlobalTracks) {
          const globalTrack = globalTracks[trackIndex];
          if (globalTrack.type === 'process') {
            hiddenPids.add(globalTrack.pid);
          }
        }

        for (const [pid, localTracks] of localTracksByPid) {
          if (hiddenPids.has(pid)) {
            // This track is hidden already.
            continue;
          }
          for (const [trackIndex, localTrack] of localTracks.entries()) {
            const hiddenLocalTracks = ensureExists(
              hiddenLocalTracksByPid.get(pid),
              'Could not look up the hidden local tracks from the given PID'
            );
            if (!hiddenLocalTracks.has(trackIndex)) {
              switch (localTrack.type) {
                case 'thread':
                  {
                    // The thread tracks have enough complexity that it warrants measuring
                    // them rather than statically using a value like the other tracks.
                    const trackThreadHeight =
                      trackThreadHeights[localTrack.threadIndex];
                    if (trackThreadHeight === undefined) {
                      // The height isn't computed yet, return.
                      return null;
                    }
                    height += trackThreadHeight + border;
                  }

                  break;
                case 'network':
                  height += TRACK_NETWORK_HEIGHT + border;
                  break;
                case 'memory':
                  height += TRACK_MEMORY_HEIGHT + border;
                  break;
                case 'event-delay':
                  height += TRACK_EVENT_DELAY_HEIGHT + border;
                  break;
                case 'ipc':
                  height += TRACK_IPC_HEIGHT + border;
                  break;
                case 'process-cpu':
                case 'power':
                  height += TRACK_PROCESS_CPU_HEIGHT + border;
                  break;
                default:
                  throw assertExhaustiveCheck(localTrack);
              }
            }
          }
        }
        return height;
      }
      default:
        throw assertExhaustiveCheck(
          timelineTrackOrganization,
          `Unhandled TimelineTrackOrganization`
        );
    }
  }
);

/**
 * This selector lets us know if it is safe to load a new profile. If
 * the app is already busy loading a profile, this selector returns
 * false.
 *
 * Used by the drag and drop component in order to determine if it can
 * load a dropped profile file.
 */
export const getIsNewProfileLoadAllowed: Selector<boolean> = createSelector(
  getView,
  getDataSource,
  getZipFileState,
  (view, dataSource, zipFileState) => {
    const appPhase = view.phase;
    const zipPhase = zipFileState.phase;
    const isLoading =
      (appPhase === 'INITIALIZING' && dataSource !== 'none') ||
      zipPhase === 'PROCESS_PROFILE_FROM_ZIP_FILE';
    return !isLoading;
  }
);

/**
 * Height of screenshot track is different depending on the view.
 */
export const getTimelineMarginLeft: Selector<number> = createSelector(
  getTimelineTrackOrganization,
  (timelineTrackOrganization) => {
    switch (timelineTrackOrganization.type) {
      case 'active-tab':
        return ACTIVE_TAB_TIMELINE_MARGIN_LEFT;
      case 'full':
      case 'origins':
        return TIMELINE_MARGIN_LEFT;
      default:
        throw assertExhaustiveCheck(
          timelineTrackOrganization,
          `Unhandled TimelineTrackOrganization`
        );
    }
  }
);

/**
 * Returns the indexes of categories that are opened in the sidebar,
 * for every category
 */
export const getSidebarOpenCategories: Selector<Map<string, Set<number>>> =
  createSelector(getApp, (app) => app.sidebarOpenCategories);
