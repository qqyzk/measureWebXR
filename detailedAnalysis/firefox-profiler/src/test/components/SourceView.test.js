/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

// @flow

import React from 'react';
import { Provider } from 'react-redux';
import { stripIndent } from 'common-tags';

import { ProfileViewer } from 'firefox-profiler/components/app/ProfileViewer';
import { updateUrlState } from 'firefox-profiler/actions/app';
import { viewProfile } from 'firefox-profiler/actions/receive-profile';
import { stateFromLocation } from 'firefox-profiler/app-logic/url-handling';
import { ensureExists } from 'firefox-profiler/utils/flow';

import {
  render,
  screen,
  within,
} from 'firefox-profiler/test/fixtures/testing-library';
import { blankStore } from 'firefox-profiler/test/fixtures/stores';
import { getProfileFromTextSamples } from 'firefox-profiler/test/fixtures/profiles/processed-profile';
import { fireFullClick } from 'firefox-profiler/test/fixtures/utils';
import { autoMockDomRect } from 'firefox-profiler/test/fixtures/mocks/domrect.js';

// We're not interested in the timeline in this test
jest.mock('../../components/timeline', () => ({
  Timeline: 'custom-timeline',
}));

describe('SourceView', () => {
  autoMockDomRect();
  afterEach(() => {
    delete Range.prototype.getClientRects;
  });

  function setup() {
    if (Range.prototype.getClientRects) {
      throw new Error(
        'jsdom now implements Range.prototype.getClientRects, please update this test.'
      );
    }

    // @codemirror/view uses getClientRects when scrolling.
    // $FlowExpectError because Flow doesn't accept assigning a function to getClientRects.
    Range.prototype.getClientRects = () => {
      return [new DOMRect(0, 0, 1024, 768)];
    };

    const revision = '997f00815e6bc28806b75448c8829f0259d2cb28';
    const filepath = 'widget/cocoa/nsAppShell.mm';
    window.fetch.get(
      `https://hg.mozilla.org/mozilla-central/raw-file/${revision}/${filepath}`,
      stripIndent`
        line 1
        line 2
        line 3
        line 4
        line 5
        line 6
        line 7
      `
    );

    const { profile } = getProfileFromTextSamples(`
      A[file:hg:hg.mozilla.org/mozilla-central:${filepath}:${revision}][line:4]
      B[file:git:github.com/rust-lang/rust:library/std/src/sys/unix/thread.rs:53cb7b09b00cbea8754ffb78e7e3cb521cb8af4b]
      C[lib:libC.so][file:s3:gecko-generated-sources:a5d3747707d6877b0e5cb0a364e3cb9fea8aa4feb6ead138952c2ba46d41045297286385f0e0470146f49403e46bd266e654dfca986de48c230f3a71c2aafed4/ipc/ipdl/PBackgroundChild.cpp:]
      D[lib:libD.so]
    `);

    const store = blankStore();
    store.dispatch(
      updateUrlState(
        stateFromLocation({
          pathname: '/from-browser',
          search: '',
          hash: '',
        })
      )
    );
    store.dispatch(viewProfile(profile));
    render(
      <Provider store={store}>
        <ProfileViewer />
      </Provider>
    );

    return {
      sourceView: () => document.querySelector('.sourceView'),
    };
  }

  it('does not show the source view at loadtime', () => {
    const { sourceView } = setup();
    expect(sourceView()).not.toBeInTheDocument();
  });

  it('should show the source view when double clicking on a line in the tree view', async () => {
    const { sourceView } = setup();

    const frameElement = screen.getByRole('treeitem', { name: /^A/ });

    fireFullClick(frameElement);
    fireFullClick(frameElement, { detail: 2 });
    expect(sourceView()).toBeInTheDocument();

    const sourceViewElement = ensureExists(sourceView());
    const sourceViewContent = await within(sourceViewElement).findByRole(
      'textbox'
    );

    // Because numbers and strings are split in several element, we're matching
    // on the string "line" only.
    await within(sourceViewContent).findAllByText('line');

    expect(sourceViewContent).toMatchSnapshot();
  });
});
