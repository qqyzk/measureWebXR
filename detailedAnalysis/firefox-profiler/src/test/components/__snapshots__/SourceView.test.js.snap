// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SourceView should show the source view when double clicking on a line in the tree view 1`] = `
<div
  aria-multiline="true"
  autocapitalize="off"
  autocorrect="off"
  class="cm-content"
  contenteditable="false"
  data-language="cpp"
  role="textbox"
  spellcheck="false"
  style="tab-size: 4;"
  tabindex="0"
  translate="no"
>
  <div
    class="cm-line"
  >
    <span
      class="tok-typeName"
    >
      line 
    </span>
    <span
      class="tok-number"
    >
      1
    </span>
  </div>
  <div
    class="cm-line"
  >
    <span
      class="tok-variableName"
    >
      line 
    </span>
    <span
      class="tok-number"
    >
      2
    </span>
  </div>
  <div
    class="cm-line"
  >
    <span
      class="tok-typeName"
    >
      line 
    </span>
    <span
      class="tok-number"
    >
      3
    </span>
  </div>
  <div
    class="cm-nonZeroLine cm-line"
  >
    <span
      class="tok-variableName"
    >
      line 
    </span>
    <span
      class="tok-number"
    >
      4
    </span>
  </div>
  <div
    class="cm-line"
  >
    <span
      class="tok-typeName"
    >
      line 
    </span>
    <span
      class="tok-number"
    >
      5
    </span>
  </div>
  <div
    class="cm-line"
  >
    <span
      class="tok-variableName"
    >
      line 
    </span>
    <span
      class="tok-number"
    >
      6
    </span>
  </div>
  <div
    class="cm-line"
  >
    <span
      class="tok-variableName"
    >
      line 
    </span>
    <span
      class="tok-number"
    >
      7
    </span>
  </div>
</div>
`;
